### ISSUE TITLE
Applying K12's "vine" construction to other sponges

### ISSUE THREAD

**Issue by DonaldTsang** [mimoo/GoKangarooTwelve#3]
Is it possible to generalize KangarooTwelve to the Vine construction ("leaves staple to a pole") for SHA3 competition round 2 candidates (reduced Fugue, Hamsi and Luffa)? And JH and CubeHash are not necessary since they are proven to be too insecure.
note: BLAKE has been modified into NORX, and Skein is a semi-sponge so there are potential as well.

**Comment by mimoo** [comment]
I would think so, also have you checked BLAKE3 which uses bao? Since it's more recent it might be better? Not sure.

**Comment by DonaldTsang** [comment]
@mimoo yeah BLAKE3 isn't sponge based but K12 is, so maybe it is possible to modify BLAKE3 into a sponge?

**Comment by mimoo** [comment]
No I meant, Bao specifically: https://www.reddit.com/r/rust/comments/a2dujv/introducing_bao_a_generalpurpose_cryptographic/

**Comment by DonaldTsang** [comment]
But bao is a tree construction that does not note on the structure of the internal hash (reduced round BLAKE which is HAIFA originally), while K12 is a Vine construction that uses a sponge function Keccak1600... 

I think that both the bao construction and the Vine construction should allow the use of ANY hash "compression" function, variety is the spice of life after all (and Vines are easier to implement than tree hashes like bao for a layman).
